src/Features/CSharp/Portable/CodeRefactorings/NamedArguments/CSharpAddNamedArgumentsCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,188 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.CodeRefactorings.NamedArguments;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Linq;
+using static Microsoft.CodeAnalysis.CSharp.SyntaxFactory;
+
+namespace Microsoft.CodeAnalysis.CSharp.CodeRefactorings.NamedArguments
+{
+    [ExportCodeRefactoringProvider(LanguageNames.CSharp, Name = nameof(CSharpAddNamedArgumentsCodeRefactoringProvider)), Shared]
+    internal class CSharpAddNamedArgumentsCodeRefactoringProvider : AbstractAddNamedArgumentsCodeRefactoringProvider<ArgumentSyntax>
+    {
+        protected override IEnumerable<ArgumentSyntax> GetNamedArguments(IList<ArgumentSyntax> arguments, ImmutableArray<IParameterSymbol> parameters)
+        {
+            if (arguments == null)
+            {
+                throw new ArgumentNullException(nameof(arguments));
+            }
+
+            for (int index = 0; index < arguments.Count; ++index)
+            {
+                var parameter = parameters[index];
+                var argument = arguments[index];
+
+                if (argument.NameColon != null)
+                {
+                    yield return argument;
+                }
+                else if (parameter.IsParams && arguments.Count - index > 1)
+                {
+                    var expressions = arguments.Skip(index).Select(arg => arg.Expression);
```

Comment:
I'm really not a fan of what's happening here.  i feel like there's a big jump between "i want named arguments" and "my params arguments are now much more involved at the callsite."  This might be another code action option.  i.e.

```
Add named arguments >
     Including 'params' arguments
     Only for literal arguments
```